src/logic/otp2/controller.ts

```diff
@@ -536,15 +513,131 @@ async function searchTaxiFrontBack(
+    // we explicitly look for trips that may include means of transport that has to be booked in advance,
+    // the second is a regular search.
+    const [
+        flexibleResults,
+        [regularTripPatterns, initialMetadata, routingErrors],
+    ] = await Promise.all([
+        searchFlexible(searchParams),
+        getTripPatterns(getTripPatternsParams),
+    ])
+
+    let queries = [
+        ...flexibleResults.queries,
+        getTripPatternsQuery(getTripPatternsParams),
+    ]
+
+    let metadata = initialMetadata
+
+    const noStopsInRangeErrors = routingErrors.filter(
+        ({ code }) => code === RoutingErrorCode.noStopsInRange,
+    )
+
+    // If we have any noStopsInRange errors, we couldn't find a means of transport from where the
+    // traveler wants to start or end the trip. Try to find an option using taxi for those parts instead.
+    let taxiTripPatterns: Otp2TripPattern[] = []
+    if (noStopsInRangeErrors.length > 0) {
+        const noFromStopInRange = noStopsInRangeErrors.some(
+            (e) => e.inputField === 'from',
+        )
+        const noToStopInRange = noStopsInRangeErrors.some(
+            (e) => e.inputField === 'to',
+        )
+
+        const taxiResults = await searchTaxiFrontBack(searchParams, {
+            access: noFromStopInRange,
+            egress: noToStopInRange,
+        })
+        taxiTripPatterns = taxiResults.tripPatterns
+        queries = [...taxiResults.queries]
+    }
+
+    let tripPatterns = regularTripPatterns
+        .filter(isValidTransitAlternative)
+        // No, this hack doesn't feel good. But we get wrong data from the backend and
+        // customers keep getting stuck on the wrong platform (31st of May 2021)
+        .map(replaceQuay1ForOsloSWithUnknown)
+
+    const nextSearchDateFromMetadata = metadata
+        ? getNextSearchDateFromMetadata(metadata, arriveBy)
+        : undefined
+
+    const flexibleTripPattern = flexibleResults.tripPatterns[0]
+    const hasFlexibleResultsOnly = flexibleTripPattern && !tripPatterns.length
+    if (hasFlexibleResultsOnly) {
+        // Flexible may return results in the future that are outside the original search window. There may still
+        // exist normal transport in the time between the search window end and the first suggested flexible result,
+        // so we do a new search to try to find those. For example, if the search window ends on midnight Friday,
+        // and the first suggested flexible result is on Monday morning, we can probably still find a normal transport
+        // option on Saturday.
+        const beforeFlexibleResult = await searchBeforeFlexible(
+            nextSearchDateFromMetadata || searchDate,
+            arriveBy,
+            flexibleTripPattern,
+            getTripPatternsParams,
+        )
+        tripPatterns = beforeFlexibleResult.tripPatterns
+        metadata = beforeFlexibleResult.metadata
+        queries = [...queries, ...beforeFlexibleResult.queries]
+    }
+
```

Comment:
This, always setting tripPatterns based on tripPatterns and something else, does not feel good.